### PR TITLE
fix for boost 1.67

### LINKED ISSUE #402: Compiling against boost-libs 1.67.0 causes ServerShutdownWithGraphicsPlatformException tests to fail
upgraded boost-libs (1.66.0-3 -> 1.67.0-4) on Arch GNU/Linux on a CI server (clean chroot environment with latest packages). All `PlatformExceptions/ServerShutdownWithGraphicsPlatformException` tests started to fail (after recompiling against 1.67.0).

Downgrading boost back to 1.66.0 and recompiling against the older version fixes the issue. I assume that 1.67.0 breaks some of the Mir tests.

Failing tests:
```
[----------] 5 tests from PlatformExceptions/ServerShutdownWithGraphicsPlatformException
[ RUN      ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/0
unknown file: Failure
C++ exception with description "std::bad_alloc" thrown in the test body.
[  FAILED  ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/0, where GetParam() = "constructor" (0 ms)
[ RUN      ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/1
unknown file: Failure
C++ exception with description "std::bad_alloc" thrown in the test body.
[  FAILED  ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/1, where GetParam() = "create_buffer_allocator" (0 ms)
[ RUN      ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/2
unknown file: Failure
C++ exception with description "std::bad_alloc" thrown in the test body.
[  FAILED  ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/2, where GetParam() = "create_display" (0 ms)
[ RUN      ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/3
unknown file: Failure
C++ exception with description "std::bad_alloc" thrown in the test body.
[  FAILED  ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/3, where GetParam() = "make_ipc_operations" (0 ms)
[ RUN      ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/4
unknown file: Failure
C++ exception with description "std::bad_alloc" thrown in the test body.
[  FAILED  ] PlatformExceptions/ServerShutdownWithGraphicsPlatformException.clean_shutdown_on_exception/4, where GetParam() = "egl_native_display" (1 ms)
[----------] 5 tests from PlatformExceptions/ServerShutdownWithGraphicsPlatformException (1 ms total)
```

### PR DESCRIPTION
**Pull request by wmww**
fixes #402

### CODE REVIEW

File: src/platform/options/program_option.cpp [hunk 1]
```diff
@@ -44,7 +44,14 @@ void mo::ProgramOption::parse_arguments(
...
     char const* argv[])
 {
-    auto parsed_command_line = po::command_line_parser(argc, argv).options(desc).allow_unregistered().run();
+    // Passing argc and argv directly to command_line_parser() should do same thing. Unfortunately, when argc is 0, boost
+    // 1.67+ does not correctly construct the vector and throws a std::bad_alloc
+    std::vector<std::string> args;
+    for (int i = 1; i < argc; i++)
+    {
+        args.push_back(*(argv + i));
+    }
```
Comment (AlanGriffiths) [inline src/platform/options/program_option.cpp]: You could avoid the loop:
```
if (argc) args.swap(std::vector{argv+1, argv+argc});
```

Comment (wmww) [reply src/platform/options/program_option.cpp]: or just `po::command_line_parser(std::max(argc, 1), argv)`

**Comment by vanyasem** [comment]
I can confirm that the issue was resolved after applying that patch

File: src/platform/options/program_option.cpp [hunk 2]
```diff
@@ -44,7 +44,10 @@ void mo::ProgramOption::parse_arguments(
     int argc,
     char const* argv[])
 {
-    auto parsed_command_line = po::command_line_parser(argc, argv).options(desc).allow_unregistered().run();
+    // Passing argc and argv directly to command_line_parser() should do same thing. Unfortunately, when argc is 0, boost
+    // 1.67+ does not correctly construct the vector and throws a std::bad_alloc
+    std::vector<std::string> args{argv + 1, argv + std::max(argc, 1)};
```
Comment (AlanGriffiths) [inline src/platform/options/program_option.cpp]: If `argc` is 0, we can't assume that `argv + 1` a valid address.

Comment (AlanGriffiths) [reply src/platform/options/program_option.cpp]: Actually, we can: argv is required to be nullptr terminated.

**Pull request merged**